primitives/core/src/offchain/storage.rs

```diff
@@ -58,3 +77,212 @@ impl OffchainStorage for InMemOffchainStorage {
 		}
 	}
 }
+
+
+
+
+/// Change to be applied to the offchain worker db in regards to a key.
+#[derive(Debug,Clone,Hash,Eq,PartialEq)]
+pub enum OffchainOverlayedChange {
+	/// Remove the data associated with the key
+	Remove,
+	/// Overwrite the value of an associated key
+	SetValue(Vec<u8>),
+}
+
+/// In-memory storage for offchain workers recoding changes for the actual offchain storage implementation.
+#[derive(Debug, Clone)]
+pub enum OffchainOverlayedChanges {
+	/// Writing overlay changes to the offchain worker database is disabled by configuration.
+	Disabled,
+	/// Overlay changes can be recorded using the inner collection of this variant.
+	Enabled(HashMap<Vec<u8>, OffchainOverlayedChange>),
+}
+
+impl Default for OffchainOverlayedChanges {
+	fn default() -> Self {
+		Self::Disabled
+	}
+}
+
+impl OffchainOverlayedChanges {
+	/// Create the disabled variant.
+	pub fn disabled() -> Self {
+		Self::Disabled
+	}
+
+	/// Create the enabled variant.
+	pub fn enabled() -> Self {
+		Self::Enabled(HashMap::new())
+	}
```

Comment:
This is not an issue but rather an opinionated comment: What is the rational for having constructors for enum? for a struct with private field I can see the reason, but here I think the raw enum syntax is just fine.

Reply:
Symmetry, if I have a constructor for one variant, I'd expect every sane user constructable variant to have one instead of mixing between enum Varaints, but this might be just my personal preference.

Reply:
>  if I have a constructor for one variant

Then the question is: why to have the constructor for the other variant in the first place? :D 

although, I can see here that `enabled` is basically `enabled_empty()`, which is a nicer syntax than explicitly building the empty HashMap.

Reply:
Because the internal representation is not relevant to the user, so I don't want to clutter outer code which has no interest in knowing about it, to construct inner structure for the overlay if that makes sense :)